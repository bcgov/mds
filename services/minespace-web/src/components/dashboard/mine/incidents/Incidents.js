// TODO: Remove this when the file is more fully implemented.
/* eslint-disable */

import React, { Component } from "react";
import { connect } from "react-redux";
import { bindActionCreators } from "redux";
import { Row, Col, Typography } from "antd";
import * as Strings from "@/constants/strings";
import PropTypes from "prop-types";
import CustomPropTypes from "@/customPropTypes";

import IncidentsTable from "@/components/dashboard/mine/incidents/IncidentsTable";
import { fetchIncidents } from "@common/actionCreators/incidentActionCreator";
import { getIncidents, getIncidentPageData } from "@common/selectors/incidentSelectors";

const { Paragraph, Title, Text } = Typography;

const propTypes = {
  fetchIncidents: PropTypes.func.isRequired,
  mine: CustomPropTypes.mine.isRequired,
  incidents: PropTypes.arrayOf(CustomPropTypes.incident).isRequired,
};

const defaultProps = {};

export class Incidents extends Component {
  state = { isLoaded: true };

  componentDidMount() {
    this.props
      .fetchIncidents({
        mine_guid: this.props.mine.mine_guid,
        per_page: Strings,
        sort_dir: "asc",
        sort_field: "mine_incident_report_no",
      })
      .then(() => {
        this.setState({ isLoaded: true });
      });
  }
  render() {
    return (
      <Row>
        <Col>
          <Title level={4}>Incidents</Title>
          <Paragraph>
            This table shows your mine's history of&nbsp;
            <Text className="color-primary" strong>
              reported incidents
            </Text>
            . The <i>Documents</i> column includes reports filed by your mine.
          </Paragraph>
          <IncidentsTable isLoaded={this.state.isLoaded} data={this.props.incidents} />
        </Col>
      </Row>
    );
  }
}

const mapStateToProps = (state) => ({
  incidentPageData: getIncidentPageData(state),
  incidents: getIncidents(state),
});

const mapDispatchToProps = (dispatch) =>
  bindActionCreators(
    {
      fetchIncidents,
    },
    dispatch
  );

Incidents.propTypes = propTypes;
Incidents.defaultProps = defaultProps;

export default connect(mapStateToProps, mapDispatchToProps)(Incidents);
