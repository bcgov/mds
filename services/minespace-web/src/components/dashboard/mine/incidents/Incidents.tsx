import React, { useContext, useEffect, useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import { useHistory } from "react-router-dom";
import { Button, Col, Row, Typography } from "antd";
import PlusCircleFilled from "@ant-design/icons/PlusCircleFilled";
import { fetchIncidents } from "@mds/common/redux/actionCreators/incidentActionCreator";
import { getIncidentPageData, getIncidents } from "@mds/common/redux/selectors/incidentSelectors";
import { DEFAULT_PAGE, DEFAULT_PER_PAGE, IMine } from "@mds/common";
import * as routes from "@/constants/routes";
import AuthorizationWrapper from "@/components/common/wrappers/AuthorizationWrapper";
import IncidentsTable from "@/components/dashboard/mine/incidents/IncidentsTable";
import { SidebarContext } from "@mds/common/components/common/SidebarWrapper";

export const Incidents = () => {
  const dispatch = useDispatch();
  const history = useHistory();
  const [isLoaded, setIsLoaded] = useState(false);
  const { mine } = useContext<{ mine: IMine }>(SidebarContext);
  const incidentPageData = useSelector(getIncidentPageData);
  const incidents = useSelector(getIncidents);

  const handleFetchIncidents = async (params?) => {
    const fetchParams = params || {
      page: DEFAULT_PAGE,
      per_page: DEFAULT_PER_PAGE,
      sort_dir: "desc",
      sort_field: "mine_incident_report_no",
    };
    await dispatch(
      fetchIncidents({
        ...fetchParams,
        mine_guid: mine.mine_guid,
      })
    );
    setIsLoaded(true);
  };

  useEffect(() => {
    handleFetchIncidents();
  }, []);

  return (
    <Row>
      <Col span={24}>
        <AuthorizationWrapper>
          <Button
            type="primary"
            className="dashboard-add-button"
            onClick={() =>
              history.push({
                pathname: routes.ADD_MINE_INCIDENT.dynamicRoute(mine.mine_guid),
                state: { mine },
              })
            }
          >
            <PlusCircleFilled />
            Record a mine incident
          </Button>
        </AuthorizationWrapper>
        <Typography.Title level={1}>Incidents</Typography.Title>
        <Typography.Paragraph>
          This table shows your mine&apos;s history of&nbsp;
          <Typography.Text className="color-primary" strong>
            reported incidents.
          </Typography.Text>
        </Typography.Paragraph>
        <Typography.Paragraph>
          <a
            href="https://www2.gov.bc.ca/gov/content/industry/mineral-exploration-mining/health-safety/incident-information"
            target="_blank"
            rel="noopener noreferrer"
          >
            Click here
          </a>{" "}
          for more information on reportable incidents.
        </Typography.Paragraph>
        {/* <IncidentsTable
          isLoaded={isLoaded}
          data={incidents}
          pageData={incidentPageData}
          handleSearch={handleFetchIncidents}
        /> */}
      </Col>
    </Row>
  );
};

export default Incidents;
