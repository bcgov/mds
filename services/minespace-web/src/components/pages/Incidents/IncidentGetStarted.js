import React from "react";
import { Col, Row, Typography, Alert } from "antd";

export const IncidentGetStarted = () => {
  return (
    <Row>
      <Col span={16} offset={4}>
        <Typography.Title level={4}>
          Verbal notice of a reportable incident, within 4 hours
        </Typography.Title>
        <Typography.Paragraph>
          Reportable incidents that cause <b>serious injury</b> or <b>loss of life</b> must be{" "}
          <b>reported verbally to the Ministry within 4 hours</b> of the incident occurrence. (HSRC
          1.7.2)
        </Typography.Paragraph>
        <Typography.Paragraph>
          Call the <b>Mine Incident Reporting Line at </b>
          <a href="tel:1-888-348-0299">
            <b>1-888-348-0299</b>
          </a>
          . This line is monitored 24/7 by an On-call Inspector.
        </Typography.Paragraph>
        <Typography.Title level={4}>
          Written notice of a reportable incident, within 16 hours
        </Typography.Title>
        <Typography.Paragraph>
          All reportable incidents must be
          <b> reported in writing to the Ministry within 16 hours</b> of the incident occurrence.
          This includes a reportable incident that caused, or had the potential to cause, serious
          injury or loss of life, or is classified as a dangerous occurrence. (HSRC 1.7.2)
        </Typography.Paragraph>
        <Typography.Paragraph strong>
          Submitting a written notice of a reportable incident within MineSpace meets your mine’s
          written reporting responsibilities under the Ministry’s Incident Management Process
          Policy.
        </Typography.Paragraph>
        <Alert
          type="warning"
          description={
            <>
              <Typography.Paragraph>
                A written notice of a reportable incident received within 4-hours satisfies both the
                requirement for the verbal report and the written report.
              </Typography.Paragraph>
            </>
          }
          showIcon
        />
        <br />
        <Typography.Title level={5}>
          Do not disturb the scene of any reportable incident
        </Typography.Title>
        <Typography.Paragraph>
          The scene of a reportable incident must not be disturbed (e.g., to return to regular
          operations), except for the purpose of <b>saving life</b> or{" "}
          <b>relieving human suffering</b>. This is to preserve evidence for both the mine’s and the
          Ministry’s investigations. (HSRC, section 1.7.3)
        </Typography.Paragraph>
        <Typography.Paragraph>
          Only your OHSC worker representative or an inspector can release the scene.
        </Typography.Paragraph>
        <Typography.Title level={5}>
          After you complete your submission in MineSpace
        </Typography.Title>
        <Typography.Paragraph>
          Mines must investigate the reportable incident and prepare an <b>investigation report</b>{" "}
          to be submitted to an Inspector of Mines (HSRC, sections 1.7.1(4) & 1.7.2).
        </Typography.Paragraph>
        <Typography.Paragraph>
          After submitting your written notice of a reportable incident through MineSpace you are
          required to come back to your submission and add your investigation report.
        </Typography.Paragraph>
        <br />
      </Col>
    </Row>
  );
};

export default IncidentGetStarted;
