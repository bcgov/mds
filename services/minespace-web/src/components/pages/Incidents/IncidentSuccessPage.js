import React from "react";
import { withRouter, Link } from "react-router-dom";
import { Row, Col, Typography, Button, Divider } from "antd";
import ArrowLeftOutlined from "@ant-design/icons/ArrowLeftOutlined";
import CheckCircleOutlined from "@ant-design/icons/CheckCircleOutlined";
import PropTypes from "prop-types";
import * as routes from "@/constants/routes";
import CustomPropTypes from "@/customPropTypes";

const propTypes = {
  location: PropTypes.shape({
    state: PropTypes.shape({
      incident: CustomPropTypes.incident,
    }),
  }).isRequired,
};

export const IncidentSuccessPage = (props) => {
  const renderContent = () => {
    const incident = props.location.state?.incident || {};
    const { mine_incident_guid, mine_guid } = incident;
    return (
      <div style={{ textAlign: "center" }}>
        <>
          <Row>
            <Col span={24}>
              <CheckCircleOutlined className="success-page" />
            </Col>
          </Row>
          <br />
          <Typography.Paragraph>
            <Typography.Title level={5}>
              Thank you, your Mines Incident has been submitted!
            </Typography.Title>
            Your submission will soon be reviewed by the ministry.
          </Typography.Paragraph>
          <Typography.Paragraph>
            This record will now move to the awaiting final investigation stage and requires you to
            visit this record again and upload your investigation documentation.
          </Typography.Paragraph>
          <div>
            <p>
              <Link to={routes.MINE_DASHBOARD.dynamicRoute(mine_guid, "incidents")}>
                <Button type="primary">Back to All incidents</Button>
              </Link>
            </p>
            <p>
              <Link
                to={{
                  pathname: routes.REVIEW_MINE_INCIDENT.dynamicRoute(mine_guid, mine_incident_guid),
                  state: { current: 2 },
                }}
              >
                <Button>View Mine Incident</Button>
              </Link>
            </p>
          </div>
        </>
      </div>
    );
  };

  return (
    <>
      <Row>
        <Col span={24}>
          <Typography.Title>
            Record a Mine Incident - {props.location.state?.incident?.mine_name || ""}
          </Typography.Title>
        </Col>
      </Row>
      <Row>
        <Col span={24}>
          <Link
            to={routes.MINE_DASHBOARD.dynamicRoute(
              props.location.state?.incident?.mine_guid,
              "incidents"
            )}
          >
            <ArrowLeftOutlined className="padding-sm--right" />
            Back to All Incidents
          </Link>
        </Col>
      </Row>
      <Divider />
      <Row>
        <Col span={24}>
          <Typography.Title level={4}>Record New Mine Incident</Typography.Title>
        </Col>
        <Col span={24}>{renderContent()}</Col>
      </Row>
    </>
  );
};

IncidentSuccessPage.propTypes = propTypes;

export default withRouter(IncidentSuccessPage);
