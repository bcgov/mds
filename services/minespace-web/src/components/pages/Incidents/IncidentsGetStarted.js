import React from "react";
import { Col, Row, Typography, Alert } from "antd";

export const IncidentsGetStarted = () => {
  return (
    <Row>
      <Col style={{ marginLeft: "15%" }}>
        <Typography.Title level={3}>Dangerous Occurence Reporting</Typography.Title>
        <Typography.Paragraph>
          You are required to report all dangerous Occurrence, pursuant to section 1.7.3 of the Code
          (including serious injuries).
        </Typography.Paragraph>
        <Typography.Title level={4}>
          Reporting a dangerous occurence through MineSpace
        </Typography.Title>
        <Typography.Paragraph>
          Mines must submit an initial report of the Dangerous Occurrence within 16 hours, pursuant
          to 1.7.1(1)(b) of the Code.
        </Typography.Paragraph>
        <Typography.Title level={4}>
          After you complete your submission in minespace
        </Typography.Title>
        <Typography.Paragraph>
          Mines must investigate the Dangerous Occurrence and prepare a report to be submitted to an
          Inspector of Mines, pursuant to 1.7.1(4) & 1.7.2 of the Code. After initial submission you
          are required to come back to your report and add your final investigation document.
        </Typography.Paragraph>
        <Typography.Title level={4}>
          Upload initial and final investigtation documents
        </Typography.Title>
        <Typography.Paragraph>
          You can upload a variety of files directly to your record. This will help ministry staff
          understand the incident and allows for faster reviews.{" "}
        </Typography.Paragraph>
        <Alert
          type="warning"
          message={<b>Did this incident result in the loss of life?</b>}
          description={
            <>
              <Typography.Paragraph>
                Immediately (within 4 hours) phone Mine Incident Reporting Line at{" "}
                <a href="tel:1-888-348-0299">1-888-348-0299</a> to submit an oral report and secure
                the accident scene.
              </Typography.Paragraph>
            </>
          }
          showIcon
        />
        <br />
      </Col>
    </Row>
  );
};

export default IncidentsGetStarted;
