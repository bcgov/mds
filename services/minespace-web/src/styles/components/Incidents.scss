.sticky-header {
  position: sticky;
  top: 80px;
  background-color: white;
  width: 100vw;
  margin-left: -42px;
  z-index: 5;
  border-bottom: rgb(194, 194,194) 2px solid;
  padding-bottom: 10px;
}

.sticky-header h1 {
  margin-bottom: 5px !important;
}

.sticky-side-menu {
  top: 220px;
  position: sticky;
  width: 240px;
}