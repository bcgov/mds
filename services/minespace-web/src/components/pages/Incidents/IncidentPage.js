import React, { Component } from "react";
import { bindActionCreators } from "redux";
import { flattenObject } from "@common/utils/helpers";
import { connect } from "react-redux";
import { Link, withRouter } from "react-router-dom";
import {
  change,
  submit,
  getFormSyncErrors,
  getFormValues,
  reset,
  touch,
  isDirty,
  destroy,
} from "redux-form";
import { Button, Row, Col, Steps, Typography, Popconfirm } from "antd";
import { ArrowLeftOutlined } from "@ant-design/icons";
import PropTypes from "prop-types";
import moment from "moment";
import { getMineIncident } from "@common/reducers/incidentReducer";
import {
  createMineIncident,
  fetchMineIncident,
  updateMineIncident,
  removeDocumentFromMineIncident,
} from "@common/actionCreators/incidentActionCreator";
import { clearMineIncident } from "@common/actions/incidentActions";
import AuthorizationGuard from "@/HOC/AuthorizationGuard";
import * as FORM from "@/constants/forms";
import * as Permission from "@/constants/permissions";
import LinkButton from "@/components/common/LinkButton";
import Loading from "@/components/common/Loading";
import customPropTypes from "@/customPropTypes";
import { IncidentGetStarted } from "@/components/pages/Incidents/IncidentGetStarted";
import IncidentForm from "@/components/Forms/incidents/IncidentForm";
import * as routes from "@/constants/routes";

const propTypes = {
  // eslint-disable-next-line react/no-unused-prop-types
  incident: customPropTypes.incident.isRequired,
  createMineIncident: PropTypes.func.isRequired,
  fetchMineIncident: PropTypes.func.isRequired,
  updateMineIncident: PropTypes.func.isRequired,
  clearMineIncident: PropTypes.func.isRequired,
  removeDocumentFromMineIncident: PropTypes.func.isRequired,
  destroy: PropTypes.func.isRequired,
  submit: PropTypes.func.isRequired,
  touch: PropTypes.func.isRequired,
  match: PropTypes.shape({
    params: PropTypes.shape({
      mineGuid: PropTypes.string,
      mineIncidentGuid: PropTypes.string,
    }),
  }).isRequired,
  location: PropTypes.shape({
    state: PropTypes.shape({
      current: PropTypes.number,
      mine: customPropTypes.mine,
    }),
  }).isRequired,
  history: PropTypes.shape({ push: PropTypes.func, replace: PropTypes.func }).isRequired,
  // eslint-disable-next-line react/no-unused-prop-types
  formValues: PropTypes.objectOf(PropTypes.any).isRequired,
  formErrors: PropTypes.objectOf(PropTypes.string),
  // eslint-disable-next-line react/no-unused-prop-types
  formIsDirty: PropTypes.bool.isRequired,
};

const defaultProps = {
  formErrors: {},
};

const StepForms = (
  props,
  state,
  navigation,
  handlers,
  formatInitialValues,
  setConfirmedSubmission,
  disabledButton
) => [
  {
    title: "Get Started",
    content: <IncidentGetStarted />,
    buttons: [
      <React.Fragment key="step-1-buttons">
        {props.formIsDirty && (
          <LinkButton
            style={{ marginRight: "15px" }}
            onClick={(e) => handlers?.save(e, props.formValues, true)}
            title="Save Draft"
          >
            Save Draft
          </LinkButton>
        )}
        {props.formIsDirty ? (
          <Popconfirm
            placement="topRight"
            title="You have unsaved work, are you sure you want to navigate away from this page?"
            okText="Yes"
            cancelText="No"
          >
            <Button
              id="step-1-cancel"
              type="secondary"
              style={{ marginRight: "15px" }}
              onClick={() =>
                props.history.push(
                  routes.MINE_DASHBOARD.dynamicRoute(props.match.params?.mineGuid, "incidents")
                )
              }
            >
              Cancel
            </Button>
          </Popconfirm>
        ) : (
          <Button
            id="step-1-cancel"
            type="secondary"
            style={{ marginRight: "15px" }}
            onClick={() =>
              props.history.push(
                routes.MINE_DASHBOARD.dynamicRoute(props.match.params?.mineGuid, "incidents")
              )
            }
          >
            Cancel
          </Button>
        )}
        <Button id="step-1-next" type="primary" onClick={() => navigation?.next()}>
          Create Record
        </Button>
      </React.Fragment>,
    ],
  },
  {
    title: "Create Record",
    content: (
      <IncidentForm
        initialValues={state.isEditMode ? formatInitialValues(props?.incident) : {}}
        handlers={{ deleteDocument: handlers?.deleteDocument }}
      />
    ),
    buttons: [
      <React.Fragment key="step-2-buttons">
        {props.formIsDirty && (
          <LinkButton
            style={{ marginRight: "15px" }}
            onClick={async (e) => {
              const response = await handlers?.save(e, props.formValues, true);
              const incidentGuid =
                props.incident?.mine_incident_guid || response?.mine_incident_guid;
              const mineGuid = props.incident?.mine_guid || response?.mine_guid;
              return props.history.push({
                pathname: `${routes.EDIT_MINE_INCIDENT.dynamicRoute(mineGuid, incidentGuid)}`,
                state: { current: 1 },
              });
            }}
            title="Save Draft"
          >
            Save Draft
          </LinkButton>
        )}
        <Button
          id="step-2-cancel"
          type="secondary"
          style={{ marginRight: "15px" }}
          onClick={() => navigation.prev()}
        >
          Back
        </Button>
        <Button
          id="step-2-next"
          type="primary"
          disabled={disabledButton}
          onClick={async (e) => {
            const response = await handlers?.save(e, props.formValues);
            const incidentGuid = props.incident?.mine_incident_guid || response?.mine_incident_guid;
            const mineGuid = props.incident?.mine_guid || response?.mine_guid;
            if (incidentGuid && mineGuid) {
              return props.history.push({
                pathname: `${routes.REVIEW_MINE_INCIDENT.dynamicRoute(mineGuid, incidentGuid)}`,
                state: { current: 2 },
              });
            }
            return null;
          }}
        >
          Review & Submit
        </Button>
      </React.Fragment>,
    ],
  },
  {
    title: "Review & Submit",
    content: (
      <IncidentForm
        setConfirmedSubmission={setConfirmedSubmission}
        confirmedSubmission={state.confirmedSubmission}
        initialValues={formatInitialValues(props?.incident)}
        isReviewSubmitStage
        applicationSubmitted={false}
      />
    ),
    buttons: [
      <React.Fragment key="step-3-buttons">
        {!props.incident.status_code && (
          <Button
            id="step-back3"
            type="tertiary"
            className="full-mobile"
            style={{ marginRight: "24px" }}
            onClick={() => {
              props.history.push({
                pathname: `${routes.EDIT_MINE_INCIDENT.dynamicRoute(
                  props.incident?.mine_guid,
                  props.incident?.mine_incident_guid
                )}`,
                state: { current: 1 },
              });
            }}
          >
            Back
          </Button>
        )}
        <Popconfirm
          placement="topRight"
          title="Are you sure you want to submit your final incident? No changes can be made after submitting."
          onConfirm={async (e) => {
            await handlers?.save(e, {
              ...props.incident,
              status_code: "IRS",
            });
            const url = routes.MINE_INCIDENT_SUCCESS.dynamicRoute(
              props.incident?.mine_guid,
              props.incident?.mine_incident_guid
            );
            const urlState = { state: { incident: props.incident } };

            return props.history.push({ pathname: url, ...urlState });
          }}
          okText="Yes"
          cancelText="No"
        >
          {!props.incident?.status_code && (
            <Button id="submit_irt" type="primary" disabled={!state.confirmedSubmission}>
              Submit Now
            </Button>
          )}
        </Popconfirm>
      </React.Fragment>,
    ],
  },
];

export class IncidentPage extends Component {
  state = {
    current: 0,
    isEditMode: false,
    isLoaded: false,
    confirmedSubmission: false,
  };

  componentDidMount() {
    this.handleFetchData().then(() => {
      this.setState((prevState) => ({
        current: this.props.location?.state?.current || prevState.current,
        isLoaded: true,
        isEditMode: Boolean(this.props.match.params?.mineIncidentGuid),
      }));
    });
  }

  componentWillUnmount() {
    this.props.clearMineIncident();
    this.props.destroy(FORM.ADD_EDIT_INCIDENT);
  }

  handleFetchData = () => {
    const { mineGuid, mineIncidentGuid } = this.props.match.params;
    if (mineGuid && mineIncidentGuid) {
      return this.props.fetchMineIncident(mineGuid, mineIncidentGuid);
    }
    return Promise.resolve();
  };

  handleCreateMineIncident = (values, isDraft) => {
    this.setState({ isLoaded: false });
    const message = isDraft ? "Successfully created a draft incident." : null;
    return this.props
      .createMineIncident(this.props.match.params?.mineGuid, values, message)
      .then((response) => {
        return response?.data;
      });
  };

  handleUpdateMineIncident = (values, isDraft) => {
    const { mineGuid, mineIncidentGuid } = this.props.match.params;
    this.setState({ isLoaded: false });

    let message;
    if (isDraft) {
      message = "Successfully updated draft incident.";
    } else if (this.props.location.state.current === 2) {
      message = "Successfully submitted a new incident.";
    } else {
      message = null;
    }

    return this.props.updateMineIncident(mineGuid, mineIncidentGuid, values, message).then(() => {
      this.handleFetchData();
      this.setState({ isLoaded: true });
    });
  };

  handleSaveData = (e, formValues, isDraft) => {
    e.preventDefault();
    this.props.submit(FORM.ADD_EDIT_INCIDENT);
    this.props.touch(FORM.ADD_EDIT_INCIDENT);
    const errors = Object.keys(flattenObject(this.props.formErrors));
    if (errors.length === 0) {
      const incidentExists = Boolean(formValues?.mine_incident_guid);
      if (!incidentExists) {
        return this.handleCreateMineIncident(this.formatPayload(formValues), isDraft);
      }
      return this.handleUpdateMineIncident(this.formatPayload(formValues), isDraft);
    }
    return null;
  };

  handleDeleteDocument = ({ mineGuid, mineIncidentGuid, mineDocumentGuid }) =>
    this.props
      .removeDocumentFromMineIncident(mineGuid, mineIncidentGuid, mineDocumentGuid)
      .then(() => this.handleFetchData());

  formatTimestamp = (dateString, time) => {
    if (!moment.isMoment(time)) {
      return dateString && time && `${dateString} ${time}`;
    }
    return dateString && time && `${dateString} ${time.format("HH:mm")}`;
  };

  formatPayload = (values) => {
    let mineDeterminationTypeCode = null;
    if (typeof values?.mine_determination_type_code === "boolean") {
      mineDeterminationTypeCode = values.mine_determination_type_code ? "DO" : "NDO";
    }
    return {
      ...values,
      categories: values?.categories?.map((cat) => cat?.mine_incident_category_code || cat),
      updated_documents: values?.initial_notification_documents,
      incident_timestamp: this.formatTimestamp(values?.incident_date, values?.incident_time),
      mine_determination_type_code:
        mineDeterminationTypeCode ?? values?.mine_determination_type_code,
    };
  };

  formatInitialValues = (incident) => ({
    ...incident,
    categories: incident?.categories?.map((cat) => cat?.mine_incident_category_code),
    incident_date: moment(incident?.incident_timestamp).format("YYYY-MM-DD"),
    incident_time: moment(incident?.incident_timestamp).format("HH:mm"),
    mine_determination_type_code: incident?.mine_determination_type_code
      ? incident.mine_determination_type_code === "DO"
      : null,
  });

  next = () => this.setState((prevState) => ({ current: prevState.current + 1 }));

  prev = () => this.setState((prevState) => ({ current: prevState.current - 1 }));

  setConfirmedSubmission = () =>
    this.setState((prevState) => ({ confirmedSubmission: !prevState.confirmedSubmission }));

  render() {
    const mineName =
      this.props.formValues?.mine_name ?? this.props.location.state?.mine?.mine_name ?? "";
    const title = `Record a Mine Incident - ${mineName}`;
    const subTitle = this.state.isEditMode ? "Edit Mine Incident" : "Record New Mine Incident";
    const errors = Object.keys(flattenObject(this.props.formErrors));
    const disabledButton = errors.length > 0;

    const Forms = StepForms(
      this.props,
      this.state,
      { next: this.next, prev: this.prev },
      { save: this.handleSaveData, deleteDocument: this.handleDeleteDocument },
      this.formatInitialValues,
      this.setConfirmedSubmission,
      disabledButton
    );

    return (
      (this.state.isLoaded && (
        <>
          <Row>
            <Col span={24}>
              <Typography.Title>{title}</Typography.Title>
            </Col>
          </Row>
          <Row>
            <Col span={24}>
              <Link
                to={routes.MINE_DASHBOARD.dynamicRoute(
                  this.props.match.params?.mineGuid,
                  "incidents"
                )}
              >
                <ArrowLeftOutlined className="padding-sm--right" />
                Back to All Incidents
              </Link>
            </Col>
          </Row>
          <br />
          <Row>
            <Col span={15}>
              <Typography.Title level={2}>{subTitle}</Typography.Title>
            </Col>
            <Col span={9}>
              <div style={{ display: "inline", float: "right" }}>
                <p>{Forms[this.state.current].buttons}</p>
              </div>
            </Col>
          </Row>
          <Row>
            <Steps current={this.state.current} style={{ marginLeft: "8%", marginRight: "8%" }}>
              {Forms.map((step) => (
                <Steps.Step key={step.title} title={step.title} />
              ))}
            </Steps>
          </Row>
          <br />
          <Row>
            <Col span={24}>
              <div>{Forms[this.state.current].content}</div>
            </Col>
          </Row>
          <Row>
            <Col span={24}>
              <div style={{ display: "inline", float: "right" }}>
                <p>{Forms[this.state.current].buttons}</p>
              </div>
            </Col>
          </Row>
        </>
      )) || <Loading />
    );
  }
}

IncidentPage.propTypes = propTypes;
IncidentPage.defaultProps = defaultProps;

const mapStateToProps = (state) => ({
  incident: getMineIncident(state) || {},
  formErrors: getFormSyncErrors(FORM.ADD_EDIT_INCIDENT)(state) || {},
  formValues: getFormValues(FORM.ADD_EDIT_INCIDENT)(state) || {},
  formIsDirty: isDirty(FORM.ADD_EDIT_INCIDENT)(state),
});

const mapDispatchToProps = (dispatch) =>
  bindActionCreators(
    {
      clearMineIncident,
      createMineIncident,
      fetchMineIncident,
      updateMineIncident,
      removeDocumentFromMineIncident,
      submit,
      reset,
      touch,
      change,
      destroy,
    },
    dispatch
  );

// ENV FLAG FOR MINE INCIDENTS //
export default withRouter(
  AuthorizationGuard(Permission.IN_TESTING)(
    connect(mapStateToProps, mapDispatchToProps)(IncidentPage)
  )
);
