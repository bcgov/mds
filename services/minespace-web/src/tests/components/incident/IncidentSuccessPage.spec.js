import React from "react";
import { shallow } from "enzyme";
import { IncidentSuccessPage } from "@/components/pages/Incidents/IncidentSuccessPage";
import * as MOCK from "@/tests/mocks/dataMocks";

const props = {};

const setupProps = () => {
  props.location = { state: { project: MOCK.INCIDENT } };
};

beforeEach(() => {
  setupProps();
});

describe("IncidentSuccessPage", () => {
  it("renders properly", () => {
    const component = shallow(<IncidentSuccessPage {...props} />);
    expect(component).toMatchSnapshot();
  });
});
