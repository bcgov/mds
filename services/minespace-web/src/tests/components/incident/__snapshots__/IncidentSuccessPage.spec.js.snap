// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IncidentSuccessPage renders properly 1`] = `
<Fragment>
  <Row>
    <Col
      span={24}
    >
      <Title>
        Record a Mine Incident - 
      </Title>
    </Col>
  </Row>
  <Row>
    <Col
      span={24}
    >
      <Link
        to="/mines/undefined/incidents?"
      >
        <ForwardRef(ArrowLeftOutlined)
          className="padding-sm--right"
        />
        Back to All Incidents
      </Link>
    </Col>
  </Row>
  <Divider />
  <Row>
    <Col
      span={24}
    >
      <Title
        level={4}
      >
        Record New Mine Incident
      </Title>
    </Col>
    <Col
      span={24}
    >
      <div
        style={
          Object {
            "textAlign": "center",
          }
        }
      >
        <Row>
          <Col
            span={24}
          >
            <ForwardRef(CheckCircleOutlined)
              className="success-page"
            />
          </Col>
        </Row>
        <br />
        <Paragraph>
          <Title
            level={5}
          >
            Thank you, your Mines Incident has been submitted!
          </Title>
          Your submission will soon be reviewed by the ministry.
        </Paragraph>
        <div>
          <p>
            <Link
              to="/mines/undefined/incidents?"
            >
              <Button
                block={false}
                ghost={false}
                htmlType="button"
                loading={false}
                type="primary"
              >
                Back to All incidents
              </Button>
            </Link>
          </p>
          <p>
            <Link
              to={
                Object {
                  "pathname": "/mines/undefined/incidents/undefined/review",
                  "state": Object {
                    "current": 2,
                  },
                }
              }
            >
              <Button
                block={false}
                ghost={false}
                htmlType="button"
                loading={false}
              >
                View Mine Incident
              </Button>
            </Link>
          </p>
        </div>
      </div>
    </Col>
  </Row>
</Fragment>
`;
